third_party/move/move-vm/runtime/src/runtime_type_checks_async.rs

```diff
@@ -0,0 +1,785 @@
+                .map_err(|err| set_err_info!(current_frame, err))?;
+            match exit {
+                ExitCode::Done => return Ok(()),
+                ExitCode::Return => {
+                    self.call_stack
+                        .type_check_return::<RTTCheck>(&mut self.stack, &mut current_frame)
+                        .map_err(|err| set_err_info!(current_frame, err))?;
+                    if let Some(frame) = self.call_stack.pop() {
+                        // Return control to the caller, advance PC past the call instruction.
+                        current_frame = frame;
+                        current_frame.pc += 1;
+                    } else {
+                        return Ok(());
+                    }
+                },
+                ExitCode::Call(idx) => {
+                    let (callee, callee_frame_cache) = self
+                        .load_function(&mut current_frame, idx)
+                        .map_err(|err| set_err_info!(current_frame, err))?;
+                    self.execute_regular_call::<RTTCheck>(
+                        &mut trace,
+                        &mut current_frame,
+                        callee,
+                        callee_frame_cache,
+                    )?;
+                },
+                ExitCode::CallGeneric(idx) => {
+                    let (callee, callee_frame_cache) = self
+                        .load_function_generic(&mut current_frame, idx)
+                        .map_err(|err| set_err_info!(current_frame, err))?;
+                    self.execute_regular_call::<RTTCheck>(
+                        &mut trace,
+                        &mut current_frame,
+                        callee,
+                        callee_frame_cache,
+                    )?;
+                },
+                ExitCode::CallClosure => {
+                    let (callee, mask) = trace.consume_closure_call().ok_or_else(|| {
+                        PartialVMError::new_invariant_violation("Call closure should be recorded")
+                            .finish(Location::Undefined)
+                    })?;
+                    let callee_frame_cache = FrameTypeCache::make_rc_for_function(&callee);
+                    self.execute_closure_call::<RTTCheck>(
+                        &mut trace,
+                        &mut current_frame,
+                        callee,
+                        callee_frame_cache,
+                        mask,
+                    )?;
+                },
+            }
+        }
+    }
+
+    /// Replays a sequence of instructions in a function frame, performing type checks. Returns an
+    /// error if checks during execution fail, or there is an internal invariant violation.
+    fn execute_instructions<RTTCheck>(
+        &mut self,
+        frame: &mut Frame,
+        trace: &mut Trace,
+    ) -> PartialVMResult<ExitCode>
+    where
+        RTTCheck: RuntimeTypeCheck,
+    {
+        loop {
```

Comment:
One way how to abstract this is to split off this loop, I'm not sure whether all those complicating, recursion-avoiding ExitCode stuff need to be specific for an abstract value domain, we just need something like `execute_bytecode`, `enter_function`, `exit_function` be implemented for a specific use case.